In [9]:
%reload_ext autoreload
%autoreload 2
%matplotlib inline

from fastai.learner import *

import torchtext
from torchtext import vocab, data
from torchtext.datasets import language_modeling

from fastai.rnn_reg import *
from fastai.rnn_train import *
from fastai.nlp import *
from fastai.lm_rnn import *

import dill as pickle



    Only loading the 'en' tokenizer.



## Language modeling

### Data

The [large movie view dataset](http://ai.stanford.edu/~amaas/data/sentiment/) contains a collection of 50,000 reviews from IMDB. The dataset contains an even number of positive and negative reviews. The authors considered only highly polarized reviews. A negative review has a score ≤ 4 out of 10, and a positive review has a score ≥ 7 out of 10. Neutral reviews are not included in the dataset. The dataset is divided into training and test sets. The training set is the same 25,000 labeled reviews.

The **sentiment classification task** consists of predicting the polarity (positive or negative) of a given text.

However, before we try to classify *sentiment*, we will simply try to create a *language model*; that is, a model that can predict the next word in a sentence. Why? Because our model first needs to understand the structure of English, before we can expect it to recognize positive vs negative sentiment.

So our plan of attack is the same as we used for Dogs v Cats: pretrain a model to do one thing (predict the next word), and fine tune it to do something else (classify sentiment).

Unfortunately, there are no good pretrained language models available to download, so we need to create our own. To follow along with this notebook, we suggest downloading the dataset from [this location](http://files.fast.ai/data/aclImdb.tgz) on files.fast.ai.

In [3]:
PATH='/home/apil/data/fastai/aclImdb_short/aclImdb/'

TRN_PATH = 'train/all/'
VAL_PATH = 'test/all/'
TRN = f'{PATH}{TRN_PATH}'
VAL = f'{PATH}{VAL_PATH}'

%ls {PATH}

imdbEr.txt  imdb.vocab  models/  README  test/  tmp/  train/


Let's look inside the training folder...

In [4]:
trn_files = !ls {TRN}
trn_files[:10]

['10014_8.txt',
 '10029_10.txt',
 '10030_0.txt',
 '10039_1.txt',
 '10057_0.txt',
 '10099_1.txt',
 '10105_0.txt',
 '10106_0.txt',
 '10110_10.txt',
 '10129_1.txt']

...and at an example review.

In [5]:
review = !cat {TRN}{trn_files[6]}
review[0]

'"The Criminal" is a noirish Brit crime flick which tells of a man (Mackintosh) who is in the wrong place at the wrong time and finds himself as a murder suspect on the run from the police and a bunch of super secret underground heavies. A busy flick with little to offer save a bunch of running around, "The Criminal" suffers from low budgetness, a thin plot, lots of filler, and an awful ending. An okay time waster for the couch potato who can\'t find anything better. (+D)'

Sounds like I'd really enjoy *Zombiegeddon*...

Now we'll check how many words are in the dataset.

In [6]:
!find {TRN} -name '*.txt' | xargs cat | wc -w

238879


In [7]:
!find {VAL} -name '*.txt' | xargs cat | wc -w

45093


Before we can analyze text, we must first *tokenize* it. This refers to the process of splitting a sentence into an array of words (or more generally, into an array of *tokens*).

In [10]:
len(spacy_tok(review[0]))

105

In [11]:
' '.join(spacy_tok(review[0]))

'" The Criminal " is a noirish Brit crime flick which tells of a man ( Mackintosh ) who is in the wrong place at the wrong time and finds himself as a murder suspect on the run from the police and a bunch of super secret underground heavies . A busy flick with little to offer save a bunch of running around , " The Criminal " suffers from low budgetness , a thin plot , lots of filler , and an awful ending . An okay time waster for the couch potato who ca n\'t find anything better . ( + D )'

We use Pytorch's [torchtext](https://github.com/pytorch/text) library to preprocess our data, telling it to use the wonderful [spacy](https://spacy.io/) library to handle tokenization.

First, we create a torchtext *field*, which describes how to preprocess a piece of text - in this case, we tell torchtext to make everything lowercase, and tokenize it with spacy.

In [12]:
TEXT = data.Field(lower=True, tokenize=spacy_tok)

fastai works closely with torchtext. We create a ModelData object for language modeling by taking advantage of `LanguageModelData`, passing it our torchtext field object, and the paths to our training, test, and validation sets. In this case, we don't have a separate test set, so we'll just use `VAL_PATH` for that too.

As well as the usual `bs` (batch size) parameter, we also not have `bptt`; this define how many words are processing at a time in each row of the mini-batch. More importantly, it defines how many 'layers' we will backprop through. Making this number higher will increase time and memory requirements, but will improve the model's ability to handle long sentences.

In [13]:
bs=64; bptt=70

In [14]:
FILES = dict(train=TRN_PATH, validation=VAL_PATH, test=VAL_PATH)
md = LanguageModelData.from_text_files(PATH, TEXT, **FILES, bs=bs, bptt=bptt, min_freq=10)

After building our `ModelData` object, it automatically fills the `TEXT` object with a very important attribute: `TEXT.vocab`. This is a *vocabulary*, which stores which words (or *tokens*) have been seen in the text, and how each word will be mapped to a unique integer id. We'll need to use this information again later, so we save it.

*(Technical note: python's standard `Pickle` library can't handle this correctly, so at the top of this notebook we used the `dill` library instead and imported it as `pickle`)*.

### Train

We have a number of parameters to set - we'll learn more about these later, but you should find these values suitable for many problems.

In [17]:
em_sz = 50  # size of each embedding vector
nh = 100     # number of hidden activations per layer
nl = 2       # number of layers

Researchers have found that large amounts of *momentum* (which we'll learn about later) don't work well with these kinds of *RNN* models, so we create a version of the *Adam* optimizer with less momentum than it's default of `0.9`.

In [18]:
opt_fn = partial(optim.Adam, betas=(0.7, 0.99))

fastai uses a variant of the state of the art [AWD LSTM Language Model](https://arxiv.org/abs/1708.02182) developed by Stephen Merity. A key feature of this model is that it provides excellent regularization through [Dropout](https://en.wikipedia.org/wiki/Convolutional_neural_network#Dropout). There is no simple way known (yet!) to find the best values of the dropout parameters below - you just have to experiment...

However, the other parameters (`alpha`, `beta`, and `clip`) shouldn't generally need tuning.

In [19]:
learner = md.get_model(opt_fn, em_sz, nh, nl,
               dropouti=0.05, dropout=0.05, wdrop=0.1, dropoute=0.02, dropouth=0.05)
learner.reg_fn = seq2seq_reg
learner.clip=0.3

As you can see below, I gradually tuned the language model in a few stages. I possibly could have trained it further (it wasn't yet overfitting), but I didn't have time to experiment more. Maybe you can see if you can train it to a better accuracy! (I used `lr_find` to find a good learning rate, but didn't save the output in this notebook. Feel free to try running it yourself now.)

In [20]:
learner.fit(3e-3, 4, wds=1e-6, cycle_len=1, cycle_mult=2)

Widget Javascript not detected.  It may not be installed or enabled properly.


[ 0.       5.76927  5.3487 ]                                
[ 1.       5.51925  5.34824]                              
[ 2.       5.45964  5.33674]                              
[ 3.       5.44803  5.33671]                              
[ 4.       5.37752  5.18741]                              
[ 5.       5.28638  5.13066]                              
[ 6.       5.23857  5.13072]                              
[ 7.       5.20148  5.04412]                              
[ 8.       5.13825  4.97651]                              
[ 9.       5.0796   4.92707]                              
[ 10.        5.03248   4.88431]                           
[ 11.        4.99687   4.86053]                           
[ 12.        4.96379   4.83377]                           
[ 13.        4.9443    4.83636]                           
[ 14.        4.93625   4.83539]                           



In [12]:
learner.save_encoder('adam1_enc')

In [21]:
learner.fit(3e-3, 4, wds=1e-6, cycle_len=10, cycle_save_name='adam3_10')

Widget Javascript not detected.  It may not be installed or enabled properly.


[ 0.       4.92743  4.81209]                              
[ 1.       4.89014  4.74529]                              
[ 2.       4.83977  4.70309]                              
[ 3.      4.8013  4.664 ]                                 
[ 4.       4.77117  4.64092]                              
[ 5.       4.74283  4.62414]                              
[ 6.       4.72666  4.61256]                              
[ 7.       4.71114  4.6073 ]                              
[ 8.       4.70529  4.60355]                              
[ 9.       4.69982  4.60734]                             
[ 10.        4.70382   4.58938]                           
[ 11.        4.68677   4.56698]                           
[ 12.        4.66374   4.54937]                           
[ 13.        4.64372   4.53684]                           
[ 14.        4.62612   4.52139]                           
[ 15.        4.61489   4.51201]                           
[ 16.        4.60185   4.50475]                          

In [14]:
learner.save_encoder('adam3_10_enc')

In [15]:
learner.fit(3e-3, 1, wds=1e-6, cycle_len=20, cycle_save_name='adam3_20')

A Jupyter Widget

[ 0.      4.2122  4.1803]                                     
[ 1.      4.4048  4.2924]                                     
[ 2.      4.4049  4.292 ]                                     
[ 3.      4.3769  4.2839]                                     
[ 4.      4.3734  4.2757]                                     
[ 5.      4.3558  4.2684]                                     
[ 6.      4.3471  4.2587]                                     
[ 7.      4.3372  4.2475]                                     
[ 8.      4.3426  4.2398]                                     
[ 9.      4.3155  4.2253]                                     
[ 10.       4.2975   4.2165]                                  
[ 11.       4.2747   4.2057]                                  
[ 12.       4.2725   4.1949]                                  
[ 13.       4.2533   4.1861]                                  
[ 14.       4.2281   4.1788]                                  
[ 15.       4.226    4.1729]                           

In [19]:
learner.load_cycle('adam3_20', 0)

In the sentiment analysis section, we'll just need half of the language model - the *encoder*, so we save that part.

In [16]:
learner.save_encoder('adam3_20_enc')

Language modeling accuracy is generally measured using the metric *perplexity*, which is simply `exp()` of the loss function we used.

In [17]:
math.exp(4.165)

64.3926824434624

### Test

We can play around with our language model a bit to check it seems to be working OK. First, let's create a short bit of text to 'prime' a set of predictions. We'll use our torchtext field to numericalize it so we can feed it to our language model.

In [62]:
m=learner.model
ss=""". So, it wasn't quite was I was expecting, but I really liked it anyway! The best"""
s = [spacy_tok(ss)]
t=TEXT.numericalize(s)
' '.join(s[0])

". So , it was n't quite was I was expecting , but I really liked it anyway ! The best"

We haven't yet added methods to make it easy to test a language model, so we'll need to manually go through the steps.

In [68]:
# Set batch size to 1
m[0].bs=1
# Turn off dropout
m.eval()
# Reset hidden state
m.reset()
# Get predictions from model
res,*_ = m(t)
# Put the batch size back to what it was
m[0].bs=bs

Let's see what the top 10 predictions were for the next word after our short text:

In [20]:
nexts = torch.topk(res[-1], 10)[1]
[TEXT.vocab.itos[o] for o in to_np(nexts)]

['performance',
 'of',
 'friend',
 'actor',
 'thing',
 'scene',
 'character',
 'part',
 'line',
 'movie']

...and let's see if our model can generate a bit more text all by itself!

In [69]:
print(ss,"\n")
for i in range(50):
    n=res[-1].topk(2)[1]
    n = n[1] if n.data[0]==0 else n[0]
    print(TEXT.vocab.itos[n.data[0]], end=' ')
    res,*_ = m(n[0].unsqueeze(0))
print('...')

. So, it wasn't quite was I was expecting, but I really liked it anyway! The best 

performance was the one in the movie where he was a little too old for the part . i think he was a good actor , but he was n't that good . 

 the movie was a bit slow , but it was n't too bad . the acting ...


### Sentiment

We'll need to the saved vocab from the language model, since we need to ensure the same words map to the same IDs.

In [70]:
TEXT = pickle.load(open(f'{PATH}models/TEXT.pkl','rb'))

`sequential=False` tells torchtext that a text field should be tokenized (in this case, we just want to store the 'positive' or 'negative' single label).

`splits` is a torchtext method that creates train, test, and validation sets. The IMDB dataset is built into torchtext, so we can take advantage of that. Take a look at `lang_model-arxiv.ipynb` to see how to define your own fastai/torchtext datasets.

In [20]:
IMDB_LABEL = data.Field(sequential=False)
splits = torchtext.datasets.IMDB.splits(TEXT, IMDB_LABEL, 'data/')

In [31]:
t = splits[0].examples[0]

In [36]:
t.label, ' '.join(t.text[:16])

('pos',
 'this was another great tom berenger movie .. but some people are right it was like')

fastai can create a ModelData object directly from torchtext splits.

In [21]:
md2 = TextData.from_splits(PATH, splits, bs)

In [43]:
m3 = md2.get_model(opt_fn, 1500, bptt, emb_sz=em_sz, n_hid=nh, n_layers=nl, 
           dropout=0.1, dropouti=0.4, wdrop=0.5, dropoute=0.05, dropouth=0.3)
m3.reg_fn = partial(seq2seq_reg, alpha=2, beta=1)
m3.load_encoder(f'adam3_20_enc')

Because we're fine-tuning a pretrained model, we'll use differential learning rates, and also increase the max gradient for clipping, to allow the SGDR to work better.

In [44]:
m3.clip=25.
lrs=np.array([1e-4,1e-3,1e-2])

In [45]:
m3.freeze_to(-1)
m3.fit(lrs/2, 1, metrics=[accuracy])
m3.unfreeze()
m3.fit(lrs, 1, metrics=[accuracy], cycle_len=1)

A Jupyter Widget

[ 0.       0.43352  0.24757  0.90585]                        



A Jupyter Widget

[ 0.       0.29882  0.1976   0.92504]                        



In [59]:
m3.fit(lrs, 7, metrics=[accuracy], cycle_len=2, cycle_save_name='imdb2')

A Jupyter Widget

[ 0.      0.275   0.1847  0.9306]                            
[ 1.      0.2528  0.1781  0.9336]                            
[ 2.      0.2448  0.1729  0.9365]                            
[ 3.      0.2052  0.1721  0.9378]                            
[ 4.      0.2239  0.1619  0.9413]                            
[ 5.      0.1956  0.1628  0.9417]                            
[ 6.      0.2093  0.158   0.9421]                            
[ 7.      0.1733  0.1591  0.9432]                            
[ 8.      0.1916  0.1552  0.9443]                            
[ 9.      0.1652  0.157   0.9451]                            
[ 10.       0.1749   0.1609   0.9427]                        
[ 11.       0.147    0.1579   0.9441]                        
[ 12.       0.1554   0.1626   0.942 ]                        
[ 13.       0.1302   0.1603   0.9437]                        
[ 14.       0.1346   0.1802   0.9384]                        
[ 15.       0.1425   0.1711   0.941 ]                        



In [78]:
m3.load_cycle('imdb2', 4)

In [79]:
accuracy(*m3.predict_with_targs())

0.94511217948717952

A recent paper from Bradbury et al, [Learned in translation: contextualized word vectors](https://einstein.ai/research/learned-in-translation-contextualized-word-vectors), has a handy summary of the latest academic research in solving this IMDB sentiment analysis problem. Many of the latest algorithms shown are tuned for this specific problem.

![image.png](attachment:image.png)

As you see, we just got a new state of the art result in sentiment analysis, decreasing the error from 5.9% to 5.5%! You should be able to get similarly world-class results on other NLP classification problems using the same basic steps.

There are many opportunities to further improve this, although we won't be able to get to them until part 2 of this course...

### End